src/mpi4py/MPI/asdlpack.pxi

```diff
@@ -151,7 +152,8 @@ cdef inline Py_ssize_t dlpack_get_itemsize(const DLTensor *dltensor) nogil:
 #------------------------------------------------------------------------------
 
 cdef int Py_CheckDLPackBuffer(object obj):
-    try: return <bint>hasattr(obj, '__dlpack__')
+    # we check __dlpack_device__ to avoid potential side effects
```

Comment:
Side effects? Now way! The `__dlpack__` attribute is a method! What side effect could possible be looking up for a method WITHOUT calling it? I think someone elsewhere suggested this approach. I don't like it and don't see the point. Are CPU-only libraries forced to expose this `__dlpack_device__` method?

Reply:
> Side effects? Now way! The `__dlpack__` attribute is a method!

You are right. This is what happens when you have a long day babysitting and try to work in midnight...😞 I forgot they are callables, not Python properties.

> Are CPU-only libraries forced to expose this `__dlpack_device__` method?

Yes. `__dlpack__` and `__dlpack_device__` must come in pair together. Libraries that implement `from_dlpack()` are expected to check `__dlpack_device__` first and act accordingly.

Reply:
OK, but then, my original code is still valid, as I'm looking up methods first, and calling them afterwards. And I would very much prefer to implement things such that `__dlpack_device__` is not mandatory.

Reply:
> I would very much prefer to implement things such that `__dlpack_device__` is not mandatory.

Unfortunately there's no agreed-upon default for the behavior of missing `__dlpack_device__` 😅 For example to CuPy defaulting to `kDLCPU` does not make sense.

Reply:
I think you should look at my code more carefully. It is true that if `__dlpack_device__` is not available, I set the `device_type` variable to `kDLCPU`. However, the only effect of that is calling `dlpack()` to get the capsule, and not `dlpack(stream=-1)`. The exporter could still expose a GPU buffer (perhaps on the default stream), and my could would still work properly. I'm going to keep insisting: the canonical way to check for DLPack suport should be to look up for the `__dlpack__` attribute, as I did originally.

Reply:
Could you check commit 4acb6bd that I just force-pushed? I reverted most of changes to `asdlpack.pxi` and now check both `__dlpack_device__` and `__dlpack__`, which I must insist.

> However, the only effect of that is calling `dlpack()` to get the capsule, and not `dlpack(stream=-1)`. The exporter could still expose a GPU buffer (perhaps on the default stream), and my could would still work properly.

I disagree that it "would still work properly". For GPU buffers, the possibility of passing `stream=None` incurs extra overhead to either set up a stream ordering or synchronize. This has to be avoided to the maximal extent allowed, as we already ask users to do this themselves. 

`__dlpack_device__` is the first function we query to obtain device info (https://github.com/data-apis/array-api/issues/142#issuecomment-796604035; see also my pseudo code in https://github.com/mpi4py/mpi4py/issues/56#issuecomment-854994872 which is followed by everyone, NumPy, CuPy, PyTorch, etc), which potentially would include more info in the future (say the API/ABI version) than a 2-tuple. If this is bothering/ambiguous to you I can send a PR to make it clear in the API standard, but if checking both methods can eliminate this corner case I don't see why we don't wanna do it.

Reply:
I think you are failing to see my point, probably because your brain is locked up in all the mess that involved putting the DLPack protocol together. Please consider the following pseudo code with the logic to get the capsule. This pseudocode is equivalent to my original implementation, without all the exception handing,

```python
dlpack = obj.__dlpack__  # required attribute
if hasattr(obj, '__dlpack_device__'): # optional attribute
    device_type, _ = obj.__dlpack_device__()
    if device_type == kDLCPU:
        capsule = dlpack()
    else:
        capsule = dlpack(stream=-1)
else:
    capsule = dlpack()
```

This implementation makes the `__dlpack_device__` attribute optional. Optional for the exporter, not for mpi4py (the consumer). mpi4py does look for `__dlpack_device__`, and if the buffer is GPU, mpi4py does call `obj.__dlpack__(stream=-1)` properly . But if, for whatever reason, the exporter library decides to not expose a `__dlpack_device__` method, my code still works and requests the DLPack capsule by calling `obj.__dlpack__()` without the stream argument. If the exporter decides to not expose the `__dlpack_device__`, well, they should be prepared for the plain call to `obj.__dlpack__()`. Why some exporter may not expose `__dlpack_device__` ? Perhaps because the exporter only deals with CPU buffers, and then `__dlpack_device__` is noisy overhead. Or perhaps the exporter handles GPU buffers always within the default stream, or always within a custom private stream, and there is no point anyway on exposing `__dlpack_device__`. Remember that the `DLTensor` metadata also contains the device type and id.

In short, I truly believe that the Python dlpack specification should NOT make `__dlpack_device__` mandatory. If they decide to do so anyway, well, I'm still free to treat `__dlpack_device__` as optional.

Can you please explain to me what is wrong with my pseudo code above? If a exporter exposes `__dlpack_device__`, I honor it, and handle the GPU path appropriately using the `stream` argument. Otherwise, I request for the capsule without a `stream` argument. In short, if a exporter handles GPUs and multiple streams, then it is the **exporter's responsibility** to expose the `__dlpack_device__` method, and mpi4py will honor it. Is my logic flawed? If yes, where? I'm just making `__dlpack_device__` optional rather than `raise AttributeError`.